tools/generators/lib/PBXProj/src/Identifiers.swift

```diff
@@ -49,13 +51,79 @@ public enum Identifiers {
     }
 
     public enum FilesAndGroups {
+        /// The logical type of the element being identified.
+        public enum ElementType: String {
+            /// A normal `PBXFileReference`.
+            case fileReference = "0"
+
+            /// A `PBXGroup`.
+            case group = "1"
+
+            /// A `PBXVariantGroup` or child `PBXFileReference`.
+            case localized = "2"
+
+            /// A `XCVersionGroup` or child `PBXFileReference`.
+            case coreData = "3"
+        }
+
         public static func mainGroup(_ path: String) -> String {
             return #"000000000000000000000003 /* \#(path) */"#
         }
 
         public static let productsGroup = #"""
 000000000000000000000004 /* Products */
 """#
+
+        /// Calculates the identifier for a file or group element at `path`.
+        ///
+        /// - Parameters:
+        ///   - path: The file path for the version group.
+        ///   - type: The type of path being identified.
+        ///   - hashCache: A cache that will be used to guarantee that the
+        ///     identifier returned is unique.
+        public static func element(
+            _ path: String,
+            type: ElementType,
+            hashCache: inout Set<String>
+        ) -> String {
+            let hash = elementHash(path + type.rawValue, hashCache: &hashCache)
+            return #"FF\#(hash) /* \#(path) */"#
```

Comment:
Currently this allows us to have 2^80 files. Which is more than enough. We might even cut away more in the future, once we determine how we want to identify targets.